### ISSUE TITLE
Add faster override for IModelExtensions.ClearIsDirtyOnAllChildren() (add suspendNotifications parameter)

### ISSUE THREAD

**Issue by Drachenkaetzchen** [Catel/Catel#1270]
`ClearIsDirtyOnAllChilds()` does not suspend change notifications, and thus fires many events. Adding a simple `using SuspendChangeNotifications()` beween [here](https://github.com/Catel/Catel/blob/95abb21dd1249b5ee5d429499895a165371d6f15/src/Catel.Core/Data/Extensions/IModelExtensions.cs#L50) and [here](https://github.com/Catel/Catel/blob/95abb21dd1249b5ee5d429499895a165371d6f15/src/Catel.Core/Data/Extensions/IModelExtensions.cs#L59) plus replacing the `IModel` cast with `ModelBase`, I was able to reduce this to 50ms.

Test prerequisites:

- ChildAwareModelBase
- 1000 childs in collection
- Bound to a ViewModel

Stock `ClearIsDirtyOnAllChilds()`: about 1000ms
With `SuspendChangeNotifications()`: about 50ms

I'm not sure what exactly the difference between `SuspendChangeNotifications()` and `SuspendChangeCallbacks()` is, but adding `SuspendChangeCallbacks()` made no measurable difference at all.

**Comment by GeertvanHorrik** [comment]
The reason is probably because you have views responding to these events? The `SuspendChangeNotifications` simply causes all notifications to be suspended (thus not updating anything).

Another "solution" could be `SetValueFastButUnsecure` (also skips notifications), but then you cannot respond to these value changes.

Change callbacks will only suspend the change callbacks *inside the model itself*, but will not suspend the `INotifyPropertyChanged` interface implementation.

So if you don't need the change notifications of the reset, you can use `SuspendChangeNotifications`. If you do need the be informed, you will need to figure out another way.

**Comment by GeertvanHorrik** [comment]
We *could* add a `suspendChangeNotification` parameter on these extension methods.

**Comment by Drachenkaetzchen** [comment]
> The reason is probably because you have views responding to these events? The `SuspendChangeNotifications` simply causes all notifications to be suspended (thus not updating anything).

Well, actually the performance impact occurs after creating the view model but before creating the view. In my use case, I load a list of `Plugin`s from the database, which are bound to a view.

As `Plugin` is a `ChildAwareModelBase`, it also responds to any changes of it's child members (which is intentional, as, for example, I need to keep track of `IsDirty` as well as the child collection counts).

> So if you don't need the change notifications of the reset, you can use `SuspendChangeNotifications`. If you do need the be informed, you will need to figure out another way.

Unfortunately yes :( I need to find another way there, as even on a "small-ish" instance of my app, the change notifications are quickly getting out of hand, especially when loading large lists from the database, in which case notifications are also fired.

I think the current model is fine for small-ish apps, but in my case, where I need to keep track of a quite large object graph in terms of object instances (but quite small in term of memory, maybe only 15-20MB) the way Catel handles it is unfortunately not going to cut it :(

**Comment by GeertvanHorrik** [comment]
I am not entirely sure what you are trying to achieve, but in our apps, we have thousands and thousands of records all implementing the `ModelBase`. It's just a matter of using it correctly (I never use `ChildAwareModelBase` or the `ClearIsDirtyOnChildren` to be honest).

**Comment by GeertvanHorrik** [comment]
What view are you loading? There are (should not) be change notifications between `ctor` & `InitializeAsync` (on the vm). It *looks* like your view has a hard time displaying large sets of data (using a datagrid?).

We used to have lots and lots of performance issues when using the datagrid, and decided to write our own from scratch (using SkiaSharp), and it's super super fast since it does not rely on WPF rendering.

**Comment by GeertvanHorrik** [comment]
Just pushed a fix. Will release an alpha (to MyGet) later today.

**Comment by Drachenkaetzchen** [comment]
> I am not entirely sure what you are trying to achieve, but in our apps, we have thousands and thousands of records all implementing the `ModelBase`. It's just a matter of using it correctly (I never use `ChildAwareModelBase` or the `ClearIsDirtyOnChildren` to be honest).

Well, let me rephrase my actual use case with a more understandable one:

Let's say I have a set of `Company` (=used to be `Plugin` in my case). Each company has a set of `Employee` (used to be `Preset` in my case). The master view which lists all companies including the count of `Employee` is displayed in a view. Additionally, I use `ChildAwareModelBase` since a user can start editing on any company, which also includes all `Employee`s. Maybe I use `ChildAwareModelBase` too lightly just to keep a per-company `IsDirty` state.

I could imagine that I'd need to implement a separate `IsDirty` state, as I could imagine that `PropertyChanged` is raised even if there's no actual change involved. I will dig a bit deeper into the issue, but right now I'm not really certain if the way I built my application is going to stay that way (especially due to having a really bad performance impact using EF6 and loading entities).

Unfortunately I'm pretty much bound to WPF DataGrid right now, as my app is completely based on Orchestra.

Thank you for the fix, I'll try it out later!

**Comment by GeertvanHorrik** [comment]
We are using Orchestra in all our apps ;-)

Anyway, you could look into light-weight entities (`ObservableObject`) to lighten your view. We are using SQLite in some apps as well, which performs really good.

In one of our apps, each entity has an `EntityState` (unchanged, added, removed, updated) and tracks its own state. Then whenever we are saving, we go through the whole object graph and submit / save any `!unchanged` state. Then we reset the state and only the really changed entities have to update in the view.

On the plugin level, you could add a `CalculatedEntityState` that is the sum of all children, but you only need to update once after saving.

**Comment by Drachenkaetzchen** [comment]
> We are using Orchestra in all our apps ;-)

:) Too bad Orchestra isn't available for OSX (yet). Many users complain that PresetMagician is Windows-only, and well...all users are musicians, so that's expected. I hope that I can use Catel together with Xamarin.Mac at some point, and I really hope to contribute some code to make that working.

> Anyway, you could look into light-weight entities (`ObservableObject`) to lighten your view. We are using SQLite in some apps as well, which performs really good.

Yes, on the basic level SQlite really performs well, unless I try to load 3k `Preset` objects for a plugin, which takes several seconds (materialization is slow). I'm currently trying to figure out why. I hope that at some point I can document some of these observations.

> In one of our apps, each entity has an `EntityState` (unchanged, added, removed, updated) and tracks its own state. Then whenever we are saving, we go through the whole object graph and submit / save any `!unchanged` state. Then we reset the state and only the really changed entities have to update in the view.

I'm actually currently trying out EF6.TrackableEntities, but I think I'll implement an own solution.
 
> On the plugin level, you could add a `CalculatedEntityState` that is the sum of all children, but you only need to update once after saving.

Yes, I gotta re-think that. Right now, I simply `RaisePropertyChanged` on the counter property. As being relatively new to C# development, performance analysis is still a big book to read ;) (being a PHP dev for 20+ years, everything is different ;))
